src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -0,0 +1,330 @@
+        links_file: LinksFile = self.links[subgraph_id]
+        links_json = links_file.content
+        entity_ids = defaultdict(set)
+        # Project ID is only mentioned in the links JSON if there is a
+        # supplementary_file_link so add it in here to make sure it is included.
+        entity_ids['project'].add(links_file.project_id)
+        link: JSON
+        for link in links_json['links']:
+            link_type = link['link_type']
+            if link_type == 'process_link':
+                entity_type = link['process_type']
+                entity_id = link['process_id']
+                entity_ids[entity_type].add(entity_id)
+                for category in ('input', 'output', 'protocol'):
+                    for file in link[f'{category}s']:
+                        entity_type = file[f'{category}_type']
+                        entity_id = file[f'{category}_id']
+                        entity_ids[entity_type].add(entity_id)
+            elif link_type == 'supplementary_file_link':
+                for file in link['files']:
+                    entity_type = file['file_type']
+                    entity_id = file['file_id']
+                    entity_ids[entity_type].add(entity_id)
+            else:
+                raise ValueError('Unknown link type', link_type)
+        return dict(entity_ids)
+
+
+@attr.s(frozen=True, kw_only=True, auto_attribs=True)
+class GitHubStagingAreaFactory:
+    repo: Repository
+    ref: str
+    path: Tuple[str, ...]
+
+    @classmethod
+    def create(cls,
+               owner: str,
+               name: str,
+               ref: str,
+               path: Optional[str] = None
+               ) -> 'GitHubStagingAreaFactory':
+        """
+        :param owner: The owner of the GitHub repository
+        :param name: The name of the GitHub repository
+        :param ref: A branch name, tag, or commit SHA
+        :param path: The path inside the repository to the base of the staging area
+        """
+        token = os.environ['GH_API_TOKEN']  # A GitHub personal access token
+        logger.debug('Requesting GitHub repo %s', (owner, name, ref, path))
+        github_api = Github(token)
+        rate_limit = github_api.get_rate_limit()
+        logger.debug('GitHub API rate limit limit=%i, remaining=%i, reset=%s',
+                     rate_limit.core.limit, rate_limit.core.remaining, rate_limit.core.reset)
+        repo = github_api.get_repo(f'{owner}/{name}')
+        path = () if path == '' else tuple(path.split('/'))
+        return cls(repo=repo, ref=ref, path=path)
+
+    @classmethod
+    def from_url(cls, url: str) -> 'GitHubStagingAreaFactory':
+        """
+        :param url: A URL of a staging area in a GitHub repository with syntax
+                    'https://github.com/<OWNER>/<NAME>/tree/<REF>[/<PATH>]'.
+                    Note that REF can be a branch, tag, or commit SHA, however
+                    if it contains '/' characters they must be URL-encoded (%2F)
```

Comment:
In documentation you should use back quotes instead of quotes as if it were markup.

```suggestion
        :param url: A URL of a staging area in a GitHub repository with syntax
                    `https://github.com/<OWNER>/<NAME>/tree/<REF>[/<PATH>]`.
                    Note that REF can be a branch, tag, or commit SHA. If REF contains special characters like `/`, '?` or `#` they must be URL-encoded. This is especially noteworthy for `/` since it's the only way to distinguish slashes in REF from those in PATH. The slashes in PATH must not be URL-encoded, while occurrences of `#` and `?` must.
```